Fix Windows arm target cross build: detect prerelease VS2017, stay in repo root

4creators opened the pull request:
Fixes issue #14514

Fixes issue #14540

Arm Windows build was recently changed by adding vswhere based VS installation detection
and earlier initialization of environment variables. For details see
https://github.com/dotnet/coreclr/pull/14228/files#diff-02389dfac0a172ab80211625efc34318R22

This introduced the following problems: (i) inability to detect and use VS 15 preview releses,
(ii) unexpected changes to working directory caused by VsDevCmd.bat script execution. These
errors prevented an arm Windows build on my machine with different errors. In particular due
to presence of arm tools only in preview version of VS the arm build consistently failed with
CMake error "compiler unknown". Fixing of this problem unmasked second bug introduced by
uncontrolled changes to build script working directory resulting in error: "Command line too
long". Both problems are fixed in this PR.

In addition in CrossArchBuild leg a new CMake custom argument is introduced which allows for
selection of latest Windows 10 SDK available on machine directly by CMake. So far this define
was present only in CoreCLR native build leg.

4creators commented:
Above changes do not provide ability to choose which VS version should be used for build and the same is true for Windows SDK versions.

PTAL
@danmosemsft @weshaggard @joperezr

weshaggard commented on build.cmd:
```diff
@@ -24,7 +25,7 @@ if defined VisualStudioVersion goto :Run
 set _VSWHERE="%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe"
 if exist %_VSWHERE% (
-  for /f "usebackq tokens=*" %%i in (`%_VSWHERE% -latest -property installationPath`) do set _VSCOMNTOOLS=%%i\Common7\Tools
+  for /f "usebackq tokens=*" %%i in (`%_VSWHERE% -latest -prerelease -property installationPath`) do set _VSCOMNTOOLS=%%i\Common7\Tools
```
I would prefer we not pass "prerelease" here as I want most folks to stay on the released builds of VS. If folks want to use a prerelease then they should simply start the build from a prerelease command prompt already.

weshaggard replied:
Just read your issue. Are you suggesting that there is no released version of the tools that will allow arm to build?

jkotas replied:
> I would prefer we not pass "prerelease" here as I want most folks to stay on the released builds of VS.

If you have installed prerelease version of VS on your machine, you may want to use it for everything. It is certainly true for folks on our team. We do want to hear about issues building with prerelease builds of VS early. The best way to achieve that is to use the prerelease build by default.

We are passing the prerelease switch here in CoreRT repo for this reason.

It matches what you get on other platforms: if you install pre-release build of XCode on Mac, we will pick it up for the build.

weshaggard replied:
Fair enough, perhaps I'm trying to be too overprotective :) Lets do it and get feedback on any issues folks hit.

4creators replied:
> they should simply start the build from a prerelease command prompt already

Problem with that approach is that starting build from prerelease command prompt will always select latest RTM installed. It is caused by prefixing vswhere with the path to standard tool which will not choose prerelease tools unless explicitly instructed to do so.

4creators replied:
>Are you suggesting that there is no released version of the tools that will allow arm to build

No, but like in my case it is possible to install VS 15 RTM without arm tools.

weshaggard replied:
>No, but like in my case it is possible to install VS 15 RTM without arm tools and then it is impossible to build for arm target since build script actively prevents using prerelease VS.

While I'm fine updating to search for prerelease as well but I'm still not sure I understand the connection of VS without arm tools and prerelease versions of VS.  In either case whatever version of VS you have installed needs to have the arm tools in order to build for arm.

4creators replied:
VS 15 RTM without arm tools when installed side by side with preview VS 15 with arm tools will always prevent using VS 15 arm tools by hijacking detection script in build.cmd and always choosing VS 15 RTM even when run from VS preview developer command prompt.

weshaggard commented on build.cmd:
```diff
@@ -314,6 +316,7 @@ REM ============================================================================
 echo %__MsgPrefix%Using environment: "%__VSToolsRoot%\VsDevCmd.bat"
```
orthogonal to your changes but I don't think this code is even needed, we should probably look closer at all the places calling vsdevcmd.bat (both in here and in run.cmd) and make them more consistent.

4creators replied:
What is your suggestion now? Should we check VsDevCmd.bat invocations in this PR or should I open new issue to track VsDevCmd.bat usage review and comment on this changes to remove them once they will not be required?

weshaggard replied:
Unless you want to don't do it as part of this PR, but having an issue would be good if you don't want to tackle it. It is just something I noticed while reviewing your change.

4creators replied:
Will do it and I have created issue #14540 to track it's progress - just assign it to me.

BruceForstall commented on build.cmd:
```diff
@@ -34,6 +35,7 @@ if not exist "%_VSCOMNTOOLS%" (
 )
 
 call "%_VSCOMNTOOLS%\VsDevCmd.bat"
+pushd %__ThisScriptDir%
```
Is there already a matching popd? (same below)

4creators replied:
No there is not. But the reason for using pushd here and below is that we have new directory pushed on directory stack by VsDevCmd.bat script. On my machine it is a default source code directory set in Visual Studio settings. Using pushd to push repo root again on top of the stack restores top of stack value to one expected by build script and changes current working directory of build script to repo root. Without that change script fails to build repo.

4creators replied:
@BruceForstall 

According to my discussion with @weshaggard above I will work on eliminating or better controlling usage of VsDevCmd.bat script in build.cmd and run.cmd what I hope will allow to get rid of unpaired `pushd`s. For now it seems to be only solution that allows to quickly fix broken build due to VsDevCmd.bat side effects.

BruceForstall replied:
I don't understand why they would ever be unpaired. It seems like you'll really mess up the user's own directory stack.

weshaggard replied:
Yeah if we cannot correctly pair it then we should probably just switch to cd instead.

BruceForstall commented on build.cmd:
```diff
@@ -484,7 +487,7 @@ if /i "%__DoCrossArchBuild%"=="1" (
     set __MsbuildWrn=/flp1:WarningsOnly;LogFile=!__BuildWrn!
     set __MsbuildErr=/flp2:ErrorsOnly;LogFile=!__BuildErr!
-
+    
```
This looks like an extraneous whitespace change

4creators replied:
Yeah, will fix it.

BruceForstall commented on build.cmd:
```diff
@@ -464,7 +467,7 @@ if /i "%__DoCrossArchBuild%"=="1" (
     set __CMakeBinDir=%__CrossComponentBinDir%
     set "__CMakeBinDir=!__CMakeBinDir:\=/!"
-    set __ExtraCmakeArgs="-DCLR_CROSS_COMPONENTS_BUILD=1" "-DCLR_CMAKE_TARGET_ARCH=%__BuildArch%" "-DCLR_CMAKE_TARGET_OS=%__BuildOs%" "-DCLR_CMAKE_PACKAGES_DIR=%__PackagesDir%" "-DCLR_CMAKE_PGO_INSTRUMENT=%__PgoInstrument%" "-DCLR_CMAKE_OPTDATA_VERSION=%__PgoOptDataVersion%" "-DCLR_CMAKE_PGO_OPTIMIZE=%__PgoOptimize%"
+    set __ExtraCmakeArgs="-DCLR_CROSS_COMPONENTS_BUILD=1" "-DCLR_CMAKE_TARGET_ARCH=%__BuildArch%" "-DCLR_CMAKE_TARGET_OS=%__BuildOs%" "-DCLR_CMAKE_PACKAGES_DIR=%__PackagesDir%" "-DCLR_CMAKE_PGO_INSTRUMENT=%__PgoInstrument%" "-DCLR_CMAKE_OPTDATA_VERSION=%__PgoOptDataVersion%" "-DCLR_CMAKE_PGO_OPTIMIZE=%__PgoOptimize%" "-DCMAKE_SYSTEM_VERSION=10.0"
```
What does `CMAKE_SYSTEM_VERSION=10.0` do? It seems strange to hard-code a version here.

4creators replied:
This change reflects already existing CMAKE_SYSTEM_VERSION=10.0 define passed to earlier call to CMake and in effect enforces that in both cases CMake will use the same SDK. According to earlier comment CMAKE_SYSTEM_VERSION=10.0 define instructs CMake to use latest SDK in 10.0.* range.

4creators commented:
Rebased to current HEAD and fixed nits. Waiting for further code review.

4creators commented:
Last commit removes unnecessary uses  of VsDevCmd.bat, fixes issue #14605 in parallel to another fix from PR #14578 but using another method - this and PR #14578 have diverged and this one should be merged first.  I am still working on #14578 due to very long test cycles ~ 1,5 hour each.

PTAL
@weshaggard @BruceForstall @jkotas

4creators commented on build-test.cmd:
```diff
@@ -120,8 +123,8 @@ if defined __BuildAgainstPackagesArg (
-echo %__MsgPrefix%Using environment: "%__VSToolsRoot%\VsDevCmd.bat"
-call                                 "%__VSToolsRoot%\VsDevCmd.bat"
+:: echo %__MsgPrefix%Using environment: "%__VSToolsRoot%\VsDevCmd.bat"
+:: call                                 "%__VSToolsRoot%\VsDevCmd.bat"
```
Commented out VsDevCmd.bat calls left intententionally to be cleared by #14578

weshaggard replied:
Why go ahead and delete them instead of commenting them out?

4creators replied:
Will fix it

4creators commented:
Ping

jkotas commented on build.cmd:
```diff
@@ -20,17 +25,21 @@ set __ThisScriptFull="%~f0"
 set _VSWHERE="%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe"
 if exist %_VSWHERE% (
-  for /f "usebackq tokens=*" %%i in (`%_VSWHERE% -latest -property installationPath`) do set _VSCOMNTOOLS=%%i\Common7\Tools
+for /f "usebackq tokens=*" %%i in (`%_VSWHERE% -latest -prerelease -property installationPath`) do set _VSCOMNTOOLS=%%i\Common7\Tools
```
This check is duplicated in several places currently. Add `-prerelease` to all of them?

4creators replied:
I agree that it is best to get it in to be consistent everywhere - anyway this is the problem which is created by code duplicates

weshaggard commented on build-test.cmd:
```diff
@@ -22,15 +26,18 @@ set __BuildOS=Windows_NT
 :: can be found.
-if defined VisualStudioVersion goto :Run
+if defined VisualStudioVersion ( 
+    if not defined __VSVersion echo %__MsgPrefix%Detected Visual Studio %VisualStudioVersion% developer command ^prompt environment
```
Why the `^`  in front of prompt?

4creators replied:
prompt is detected as keyword - ^ put in front escapes it and is not visible when printed out

weshaggard commented on tests/runtest.cmd:
```diff
@@ -22,15 +26,18 @@ set __MSBuildBuildArch=x64
 :: can be found.
-if defined VisualStudioVersion goto :Run
+if defined VisualStudioVersion ( 
+    if not defined __VSVersion echo %__MsgPrefix%Detected Visual Studio %VisualStudioVersion% developer command ^prompt environment
```
What is the use of this message? We should still default __VSVersion from the VisualStudioVersion similar to https://github.com/dotnet/corefx/blob/master/src/Native/build-native.cmd#L69.

4creators replied:
I wanted to have a message which tests for consistency of VS developer command prompt detection and prints result. One of the reasons for some builds failing on my machine was a problem with VS environment detection and unnecessary execution of VsDevCmd.bat for preview VS. Now it is fixed and reports version detected.

In all scripts there is a new block for detecting VS installation and it's version introduced in #14228. I have not changed it - if it should be changed I can do it just let me know.

`if not defined __VSVersion echo` usage is due to the fact that currently multiple scripts contain this detection code and this conditional eliminates duplicate messages from any script called from a top level one.

weshaggard replied:
I'm good with the message for now, lets handle cleaning this up and making it consistent as part of the other issues.

weshaggard commented on tests/runtest.cmd:
```diff
@@ -154,11 +157,11 @@ if not exist "%__VCToolsRoot%\vcvarsall.bat"          goto NoVS
 :: Set the environment for the  build- VS cmd prompt
-echo %__MsgPrefix%Using environment: "%__VSToolsRoot%\VsDevCmd.bat"
-call                                 "%__VSToolsRoot%\VsDevCmd.bat"
+:: echo %__MsgPrefix%Using environment: "%__VSToolsRoot%\VsDevCmd.bat"
```
if these aren't needed we should just delete instead of comment out.

4creators replied:
Will fix it

4creators commented:
Fixed unmatching `pushd` and `popd` - they are called from other scripts  and after logging script execution it appears that originally one `pushd` was missing. Therefore I have deleted one of two `pushd` added and other was left to keep directory stack balanced.

4creators commented:
Addressed code review feedback, left comments on issues in code to decide on, CI is green.
PTAL cc @jkotas @weshaggard

weshaggard commented:
Thanks for helping improve the build scripts @4creators!!!
